File: app/sample-viewmodel/src/main/java/com/open/viewmodel/test/asasa/MainActivity.java
package com.open.viewmodel.test.asasa;

import android.app.Activity;
import android.app.Instrumentation;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.open.viewmodel.R;

import java.lang.reflect.Field;

/**
 * MainActivity
 *
 * @Description: xxx
 * @Author: xing.tang
 */
public class MainActivity extends AppCompatActivity {

    @Override
    protected void attachBaseContext(Context newBase) {
        super.attachBaseContext(newBase);
        // replaceActivityInstrumentation
        replaceActivityInstrumentation(this);
        // replaceContextInstrumentation
        replaceContextInstrumentation(this);
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        // replaceActivityInstrumentation
        Intent intent1 = new Intent(Intent.ACTION_VIEW);
        intent1.setData(Uri.parse("http://zhangsan.com"));
        startActivity(intent1);
        // replaceContextInstrumentation
        Intent intent2 = new Intent(Intent.ACTION_VIEW);
        intent2.setData(Uri.parse("http://zhangsan.com"));
        intent2.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        getApplicationContext().startActivity(intent2);
    }



    public void replaceActivityInstrumentation(Activity activity){
        try{
            // 得到 Activity 的 mInstrumentation 字段
            Field field = Activity.class.getDeclaredField("mInstrumentation");// 1
            // 取消 Java 的权限控制检查
            field.setAccessible(true);// 2
            Instrumentation instrumentation = (Instrumentation) field.get(activity);// 3
            Instrumentation instrumentationProxy = new InstrumentationProxy(instrumentation);// 4
            field.set(activity,instrumentationProxy);
        }catch (Exception e){
            e.printStackTrace();
        }
    }

    public void replaceContextInstrumentation(Activity activity){
        try{
            // 获取 ActivityThread 类
            Class<?> activityThreadClazz =  Class.forName("android.app.ActivityThread");
            Field activityThreadField = activityThreadClazz.getDeclaredField("sCurrentActivityThread");// 1
            activityThreadField.setAccessible(true);
            Object currentActivityThread = activityThreadField.get(null);// 2
            // 获取 ActivityThread 中的 mInstrumentation 字段
            Field mInstrumentationField = activityThreadClazz.getDeclaredField("mInstrumentation");
            mInstrumentationField.setAccessible(true);
            Instrumentation instrumentation = (Instrumentation) mInstrumentationField.get(currentActivityThread);
            Instrumentation instrumentationProxy = new InstrumentationProxy(instrumentation);// 3
            mInstrumentationField.set(currentActivityThread,mInstrumentationField);
        }catch (Exception e){
            e.printStackTrace();
        }
    }
}

File: app/sample-viewmodel/src/main/java/com/open/viewmodel/MainActivity.kt
package com.open.viewmodel

import android.app.Activity
import android.app.Instrumentation
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.lifecycle.SavedStateViewModelFactory
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.observe
import com.open.common.base.BaseActivity
import com.open.viewmodel.databinding.ActivityMainBinding
import com.open.viewmodel.test.asasa.InstrumentationProxy
import ws.vinta.pangu.Pangu

class MainActivity : BaseActivity<ActivityMainBinding>() {

    override fun isOpenARouter(): Boolean = false

    // private val mViewModel: TestModel by viewModels()
    private lateinit var testViewModel: TestModel

    private lateinit var savedstateViewModel: SavedstateViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        replaceActivityInstrumentation(this)

        testViewModel = ViewModelProvider(this)[TestModel::class.java]
        testViewModel.testLiveData.observe(this) {
            mViewBinding.tvViewmodelText.text = it.toString()
        }

        savedstateViewModel = ViewModelProvider(
            this,
            SavedStateViewModelFactory(application, this)
        )[SavedstateViewModel::class.java]
        savedstateViewModel.testLiveData.observe(this) {
            mViewBinding.tvViewmodelText.text = it.toString()
        }
    }

    fun onClick(v: View) {
        when (v.id) {
            R.id.btn_viewmodel_loadinit -> {
                testViewModel.loadInitData()

                val pangu = Pangu()
                val newText = pangu.spacingText("請問Jackie的鼻子有幾個？123個！")
                Log.d("MainActivity", "onCreate: $newText")
            }
            R.id.btn_viewmodel_loadinit_by_savedstate -> {
                savedstateViewModel.loadInitData()
                val intent = Intent(Intent.ACTION_VIEW)
                intent.data = Uri.parse("http://zhangsan.com")
                startActivity(intent)
            }
        }
    }

    fun replaceActivityInstrumentation(activity: Activity?) {
        try {
            // 得到 Activity 的 mInstrumentation 字段
            val field = Activity::class.java.getDeclaredField("mInstrumentation") // 1
            // 取消 Java 的权限控制检查
            field.isAccessible = true // 2
            val instrumentation = field[activity] as Instrumentation // 3
            val instrumentationProxy: Instrumentation = InstrumentationProxy(instrumentation) // 4
            field[activity] = instrumentationProxy
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}